docs/design/2022-03-25-non-transactional-statement.md

```diff
@@ -0,0 +1,123 @@
+
+## Introduction
+
+A new syntax that transform a given DML into a serial execution of mutually exclusive collectively exhaustive statements. 
+
+For example, `split on a limit 1000 delete from t` can be transformed into `delete from t where a between 1 and 1000` and `delete from t where a between 1001 and 2000`, assuming the min and max of `a` are respectively 1 and 2000.
+
+## Motivation or Background
+
+Users need to do a bulk delete or update using a single statement, while currently TiDB cannot satisfy the requirement because of the performance issue, transaction size limit, and compatibility issues with tools.
+
+A new syntax is proposed to work around the problem. A non-transactional DML contains a DML and information that are used to "split" the statement, thus it is equivalent to a sequence of DMLs which is not transactional since it does not provide atomicity and probably isolation.
+
+Different from the deprecated batch-DML, a non-transactional DML splits SQLs so every split SQL is a normal statement and does not risk data-index consistency.
+
+## Detailed Design
+
+### Syntax
+
+The syntax: `split on <column_name> limit <batch_size> <DML>`
+
+In the first step only `delete` is going to be supported, but `update` and `insert into select` are also worth considering.
+
+There is no other constraints on the DML.
+
+There can be a dry run syntax `dryrun split on <column_name> limit <batch_size> <DML>` that just output how the statement will be split and an representative of their query plans.
+
+### Session
+
+Different from most of the statements, the non-transactional DML is handled in the session level. 
+
+Non-transactional statements are treated as `SimplePlans` and we won't compile it. In `runStmt`, instead of `s.Exec` for most plans, `handleNontransactionalDML` will be called. In `handleNontransactionalDML`, we will split the DML into multiple statements and execute them one by one using `Session.ExecuteStatement`.
+
+### How the split work
+
+To find the split keys, a `select` is used to read the split column specified by the user. For example, for `split on a limit 1000 delete from t where b < 1000`, the select statement would look like: `select a from t where b < 1000 order by a`.
+The result set could be large, but we don't need all of them, so result chunks are batched until the size of the batch is greater than or equal to the specified `batchSize`. Only the first and last element of each batch are kept, and thus forming a job. A job specifies a range in the specified column.
+
+Theoretically jobs can be executed in parallel, but that would require multiple sessions to execute them. There are two kinds of sessions in TiDB: user sessions that must be bound to client connections in a 1-on-1 manner, and internal sessions. Both kinds of sessions are inappropriate the split statements, so in favor of maintainence, only the current user session will be used to execute the jobs, which is serial.
```

Comment:
One of the problems is not enough concurrency when using a single large transaction. After adding this limit, I am a bit afraid how much performance edge there will be using split transactions.

Reply:
Yes. It's trading performance for maintenance. Serial execution is acceptable for some users.

Reply:
Better to clarify this, the user may compare the performance of splitted queries with normal ones, and some batch executions must finish in the night before the next day.